### PR TITLE
Get rid of System from ThreadedElementLoop

### PR DESCRIPTION
**Pull request by fdkong**
`ThreadedElementLoop` does not necessarily depend on
`System`.

Refs #7398 #7654

### CODE REVIEW

**Comment by andrsd** [comment]
In general, this is ok, but can be even improved in a following way. When we pass in `FEProblem` and `NonlinearSystem` as parameters in the c-tor, we actually need just the `FEProblem`. The instance of `NonlinearSystem` can be obtained in the initializer list as `_nl(_fe_problem.getNonlinearSystem()),`.

Similarly for `AuxiliarySystem`...

**Comment by friedmud** [comment]
Ummm... the whole purpose of `ThreadedElementLoopBase` is to be a "threaded element loop" without a System.  Just use that instead if you need to do that.

**Comment by fdkong** [comment]
> In general, this is ok, but can be even improved in a following way. When we pass in FEProblem and NonlinearSystem as parameters in the c-tor, 

For which classes? For any class which has a constructor like `Class(FEProblem &, NonlinearSystem/AuxiliarySystem &)` ?

>we actually need just the FEProblem. The instance of NonlinearSystem can be obtained in the initializer list as _nl(_fe_problem.getNonlinearSystem()),.
> 
> Similarly for AuxiliarySystem...

**Comment by andrsd** [comment]
> For any class which has a constructor like Class(FEProblem &, NonlinearSystem/AuxiliarySystem &) ?

Yes.

**Comment by fdkong** [comment]
> Ummm... the whole purpose of ThreadedElementLoopBase is to be a "threaded element loop" without a System. Just use that instead if you need to do that.

I don't  see any good reason to let `ThreadedElementLoopBase` depend on `System` either.

The dependency also causes some issues. Please see comments  we have at #7654.

**Comment by fdkong** [comment]
> For any class which has a constructor like Class(FEProblem &, NonlinearSystem/AuxiliarySystem &) ?
> 
> Yes.

There a lot of classes like this. Let me improve this in another PR.

**Comment by andrsd** [comment]
> Let me improve this in another PR.

Why another PR? Another patch within this PR is going to be just fine...

**Comment by fdkong** [comment]
> Why another PR? Another patch within this PR is going to be just fine...

I am fine with this. I just won't give reviewers have hard time

**Comment by friedmud** [comment]
`ThreadedElementLoopBase` does _not_ depend on `System`.  That is the point of that class.

**Comment by andrsd** [comment]
[This](https://github.com/idaholab/moose/pull/7654#issuecomment-248159650) is the motivation behind this change. The point is that not even the child classes need `System` passed in via c-tor...

**Comment by friedmud** [comment]
Ok - so are you going to remove `ThreadedElementLoopBase`?

**Comment by andrsd** [comment]
I did not have that in mind, but it looks like it should be possible...

**Comment by fdkong** [comment]
> Ok - so are you going to remove ThreadedElementLoopBase?

Possibly not. `ThreadedElementLoopBase` depends on `MooseMesh` only but `ThreadedElementLoop` also depends on `FEProblem` for setting an exception something like that.

**Comment by andrsd** [comment]
I think that we never loop over the displaced version of the mesh, so having `FEProblem` gives us the right `MooseMesh` that the `ThreadedElementLoopBase` needs right now. Thus, it should be possible to merge the `ThreadedElementLoopBase` into `ThreadedElementLoop`...

**Comment by fdkong** [comment]
> I think that we never loop over the displaced version of the mesh, so having FEProblem gives us the right MooseMesh that the ThreadedElementLoopBase needs right now. Thus, it should be possible to merge the ThreadedElementLoopBase into ThreadedElementLoop...

Do you want this in the same PR too, @andrsd ?

**Comment by andrsd** [comment]
> Do you want this in the same PR too, @andrsd ?

Yes.

**Comment by fdkong** [comment]
@YaqiWang's comments disappeared after I did something like:

`git commit -a --amend `

`git push -f origin get_rid_of_system_from_ThreadedElementLoop`

**Comment by fdkong** [comment]
@andrsd, `CacheChangedListsThread` inherits from `ThreadedElementLoopBase` directly.  `CacheChangedListsThread` is used in `MooseMesh` where we do not have access to `FEProblem` that is supposed as an input parameter for `ThreadedElementLoop`. 

If we merge `ThreadedElementLoopBase` into `ThreadedElementLoop`, then `CacheChangedListsThread` has to derive from `ThreadedElementLoop` instead of `ThreadedElementLoopBase`. In `MooseMesh`, we have no way to pass `FEProblem` to  `CacheChangedListsThread`.

**Comment by andrsd** [comment]
Ok

**Comment by fdkong** [comment]
> Ok

Is this PR ready to go? Or you need more time to review again?

**Review by YaqiWang — APPROVED** [review]

**Comment by YaqiWang** [comment]
To me it is ok. But framework developers may have more comments.

**Comment by fdkong** [comment]
> To me it is ok. But framework developers may have more comments.

@idaholab/moose-team, any comments?

**Review by permcody — COMMENTED** [review]
This looks good to me 👍

**Comment by permcody** [comment]
@andrsd Please merge since you've been guiding @fdkong on this change.

**Comment by andrsd** [comment]
I am on it...

File: framework/include/base/ComputeFullJacobianThread.h [hunk 1]
```diff
@@ -39,6 +39,9 @@ class ComputeFullJacobianThread : public ComputeJacobianThread
   virtual void computeInternalFaceJacobian(const Elem * neighbor) override;
   virtual void computeInternalInterFaceJacobian(BoundaryID bnd_id) override;
 
+  // Reference to NonlinearSystem
```
Comment (andrsd) [inline framework/include/base/ComputeFullJacobianThread.h]: Remove this line, this is cpt. Obvious...

Comment (fdkong) [inline framework/include/base/ComputeFullJacobianThread.h]: Done

**Review by andrsd — APPROVED** [review]

**Pull request merged**